agent/php_user_instrument_wraprec_hashmap.c

```diff
@@ -0,0 +1,482 @@
+  bucket->key->name_len = key->name_len;
+  bucket->key->name_hash = key->name_hash;
+  bucket->wraprec = nr_php_user_wraprec_create();
+
+  if (hashmap->buckets[hash_key]) {
+    hashmap->buckets[hash_key]->prev = bucket;
+  }
+
+  hashmap->buckets[hash_key] = bucket;
+  ++hashmap->elements;
+
+  return bucket->wraprec;
+}
+
+static nruserfn_t* nr_func_hashmap_lookup_internal(nr_func_hashmap_t* hashmap, nr_func_hashmap_key_t* key) {
+  size_t hash;
+  nr_func_bucket_t* bucket;
+
+  if (nrunlikely((NULL == hashmap) || (NULL == key))) {
+    return NULL;
+  }
+
+  hash = nr_func_hashmap_hash_key(hashmap->log2_num_buckets, key);
+  if (nr_func_hashmap_fetch_internal(hashmap, hash, key, &bucket)) {
+    return bucket->wraprec;
+  }
+
+  return NULL;
+}
+
+static nruserfn_t* nr_func_hashmap_update_internal(nr_func_hashmap_t* hashmap, nr_func_hashmap_key_t* key, bool* created) {
+  size_t hash;
+  nr_func_bucket_t* bucket;
+  nruserfn_t* wraprec;
+
+  if (nrunlikely((NULL == hashmap) || (NULL == key))) {
+    return NULL;
+  }
+
+  hash = nr_func_hashmap_hash_key(hashmap->log2_num_buckets, key);
+  if (nr_func_hashmap_fetch_internal(hashmap, hash, key, &bucket)) {
+    if (created) {
+      *created = false;
+    }
+    return bucket->wraprec;
+  }
+
+  if (created) {
+    *created = true;
+  }
+  return nr_func_hashmap_add_internal(hashmap, hash, key);
+}
+
+static void nr_func_hashmap_destroy_bucket_internal(nr_func_bucket_t** bucket_ptr) {
+  nr_func_bucket_t* bucket = *bucket_ptr;
+  nr_free(bucket->key->name);
+  nr_free(bucket->key);
+  nr_php_user_wraprec_destroy(&bucket->wraprec);
+  nr_realfree((void**)bucket_ptr);
+}
+
+static void nr_func_hashmap_destroy_internal(nr_func_hashmap_t** hashmap_ptr) {
+  size_t count;
+  nr_func_hashmap_t* hashmap;
+  size_t i;
+
+  if ((NULL == hashmap_ptr) || (NULL == *hashmap_ptr)) {
+    return;
+  }
+  hashmap = *hashmap_ptr;
+
+  count = (size_t)(1 << hashmap->log2_num_buckets);
+  for (i = 0; i < count; i++) {
+    nr_func_bucket_t* bucket = hashmap->buckets[i];
+
+    while (bucket) {
+      nr_func_bucket_t* next = bucket->next;
+
+      nr_func_hashmap_destroy_bucket_internal(&bucket);
+
+      bucket = next;
+    }
+  }
+
+  nr_free(hashmap->buckets);
+  nr_realfree((void**)hashmap_ptr);
+}
+
+// -----------------------------------------------------------------------------
+// scope hash map
+
+typedef struct {
+  const char* name;
+  int name_len;
+  zend_ulong name_hash;
+} nr_scope_hashmap_key_t;
+
+typedef struct _nr_scope_bucket {
+  struct _nr_scope_bucket* prev;
+  struct _nr_scope_bucket* next;
+  nr_scope_hashmap_key_t* key;
+  nr_func_hashmap_t* scoped_funcs_ht;
+} nr_scope_bucket_t;
+
+typedef struct _nr_scope_hashmap {
+  size_t log2_num_buckets;
+  nr_scope_bucket_t** buckets;
+  size_t elements;
+} nr_scope_hashmap_t;
+
+static nr_scope_hashmap_t* nr_scope_hashmap_create_internal(size_t log2_num_buckets) {
+  nr_scope_hashmap_t* hashmap;
```

Comment:
```suggestion
  nr_scope_hashmap_t* hashmap = NULL;
```

General comment regarding the other unitialized pointers in the PR, it's good practice to initialize the pointers to NULL.

Reply:
See 55ea3c5a1fdf2bc65e4df93f9a2f65db0ed40747.